bug: json is not escaped correctly in the update workflow

This was raised by @BigLep 
https://filecoinproject.slack.com/archives/C03KLC57LKB/p1659909313576869

The update workflow keeps crashing with

```
parse error: Invalid numeric literal at line 1, column 16107
Error: Process completed with exit code 4.
```

- workflow: https://github.com/ipfs/github-mgmt/actions/workflows/update.yml
- example of error: https://github.com/ipfs/github-mgmt/actions/runs/2813488216

This is caused by the quotes in the title of https://github.com/ipfs/github-mgmt/pull/24

Escaping is lost before we enter the loop body in update.yml, and the `"` in the title breaks the JSON.
We execute something like:

```sh
pr='{"title": "hello"woorld"", "number": 12"}' 
$(jq -r '.number' <<< "$pr")
```

--- a/.github/workflows/update.yml
+++ b/.github/workflows/update.yml
@@ -20,7 +20,7 @@ jobs:
       run:
         shell: bash
     steps:
-      - uses: actions/checkout@v2
+      - uses: actions/checkout@v3
       - run: npm install && npm run build
         working-directory: scripts
       - name: Update PRs

--- a/.github/workflows/update.yml
+++ b/.github/workflows/update.yml
@@ -20,40 +20,13 @@ jobs:
       run:
         shell: bash
     steps:
-      - name: Generate app token
-        id: token
-        uses: tibdex/github-app-token@7ce9ffdcdeb2ba82b01b51d6584a6a85872336d4 # v1.5.1
-        with:
-          app_id: ${{ secrets.RW_GITHUB_APP_ID }}
-          installation_id: ${{ secrets[format('RW_GITHUB_APP_INSTALLATION_ID_{0}', github.repository_owner)] }}
-          private_key: ${{ secrets.RW_GITHUB_APP_PEM_FILE }}
+      - uses: actions/checkout@v2
+      - run: npm install && npm run build
+        working-directory: scripts
       - name: Update PRs
         env:
-          GITHUB_TOKEN: ${{ steps.token.outputs.token }}
-        run: |
-          while read pr; do
-            if [[ ! -z "$pr" ]]; then
-              number="$(jq -r '.number' <<< "$pr")"
-              user_type="$(jq -r '.user.type' <<< "$pr")"
-              if [[ "$user_type" == 'Bot' ]]; then
-                echo '{"message":"Not updating pull request branch. The pull request was created by a bot user.","url":"https://api.github.com/repos/${{ github.repository }}/pulls/'"$number"'"}'
-                continue
-              fi
-              private="$(jq -r '.head.repo.private' <<< "$pr")"
-              if [[ "$private" == 'false' ]]; then
-                base_label="$(jq -r '.base.label' <<< "$pr")"
-                head_label="$(jq -r '.head.label' <<< "$pr")"
-                behind_by="$(gh api "/repos/${{ github.repository }}/compare/$base_label...$head_label" --jq '.behind_by')"
-                if [[ "$behind_by" == '0' ]]; then
-                  echo '{"message":"Not updating pull request branch. Head repository is up to date.","url":"https://api.github.com/repos/${{ github.repository }}/pulls/'"$number"'"}'
-                  continue
-                fi
-              fi
-              if ! gh api -X PUT "/repos/${{ github.repository }}/pulls/$number/update-branch"; then
-                echo ''
-                echo "::warning title=update-branch failure::${{github.repository}}#$number"
-              else
-                echo ''
-              fi
-            fi
-          done <<< "$(gh api '/repos/${{ github.repository }}/pulls' --jq '.[]')"
+          GITHUB_APP_ID: ${{ secrets.RW_GITHUB_APP_ID }}
+          GITHUB_APP_INSTALLATION_ID: ${{ secrets[format('RW_GITHUB_APP_INSTALLATION_ID_{0}', matrix.workspace)] || secrets.RW_GITHUB_APP_INSTALLATION_ID }}
+          GITHUB_APP_PEM_FILE: ${{ secrets.RW_GITHUB_APP_PEM_FILE }}
+        run: node lib/actions/update-pull-requests.js
+        working-directory: scripts

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -29,3 +29,4 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### Fixed
 - links to supported resources in HOWTOs
 - posting PR comments when terraform plan output is very long
+- PR parsing in the update workflow

--- a/scripts/package.json
+++ b/scripts/package.json
@@ -17,6 +17,7 @@
   "dependencies": {
     "@actions/core": "^1.9.0",
     "@actions/exec": "^1.1.1",
+    "@actions/github": "^5.0.3",
     "@octokit/auth-app": "^4.0.2",
     "@octokit/graphql": "^4.8.0",
     "@octokit/plugin-retry": "^3.0.9",

--- a/scripts/src/actions/update-pull-requests.ts
+++ b/scripts/src/actions/update-pull-requests.ts
@@ -1,0 +1,41 @@
+import {GitHub} from '../github'
+import {context} from '@actions/github'
+
+async function updatePullRequests() {
+  const github = await GitHub.getGitHub()
+
+  const pulls = await github.client.paginate(github.client.pulls.list, {
+    ...context.repo,
+    state: 'open'
+  })
+
+  for (const pull of pulls) {
+    if (pull.draft === true) {
+      // skip draft pull requests
+      continue
+    }
+
+    if (pull.user?.type === 'Bot') {
+      // skip bot pull requests
+      continue
+    }
+
+    // replace process.env.GITHUB_REF_NAME with context.refName if it becomes available https://github.com/actions/toolkit/pull/935
+    if (pull.base.ref !== (process.env.GITHUB_REF_NAME as string)) {
+      // skip pull requests that are not on the target branch
+      continue
+    }
+
+    if (pull.base.sha === context.sha) {
+      // skip pull requests that are already up to date
+      continue
+    }
+
+    github.client.pulls.updateBranch({
+      ...context.repo,
+      pull_number: pull.number
+    })
+  }
+}
+
+updatePullRequests()

